elements/howto-label/howto-label.js

```diff
@@ -0,0 +1,151 @@
+      // so it makes sense to cache a reference to it.
+      this._slot = this.shadowRoot.querySelector('slot');
+      // Listen for the slotchange event to label any wrapped children
+      this._slot.addEventListener('slotchange', this._updateLabel.bind(this));
+      this.addEventListener('click', this._onClick.bind(this));
+    }
+
+    /**
+     * connectedCallback() will run once the element is attached to the document.
+     * It's the earliest point at which you may touch attributes, and is a good
+     * time to auto-generate an id for the element if it does not already have
+     * one. The labelled element will use this id to refer to the label with
+     * aria-labelledby.
+     */
+    connectedCallback() {
+      // Fire _updateLabel at startup to label any wrapped elements.
+      this._updateLabel();
+    }
+
+    get for() {
+      const value = this.getAttribute('for');
+      return value === null ? '' : value;
+    }
+
+    set for(value) {
+      this.setAttribute('for', value);
+    }
+
+    attributeChangedCallback(name, oldVal, newVal) {
+      this._updateLabel();
+    }
+
+    /**
+     * Remove any previous labels.
+     * Find the new element to label.
+     * Label it.
+     */
+    _updateLabel() {
+      // Greedily generate id if one is not already present.
+      if (!this.id) {
+        this.id = `howto-label-generated-${howtoLabelCounter++}`;
+      }
+      let oldTarget = this._currentLabelTarget();
+      let newTarget = this._findTarget();
+      if (oldTarget === newTarget) {
+        return;
+      }
+      if (oldTarget) {
+        oldTarget.removeAttribute('aria-labelledby');
+      }
+      newTarget.setAttribute('aria-labelledby', this.id);
+    }
+
+    _onClick(event) {
+      let el = this._currentLabelTarget();
+      // If nothing is labeled, or if a wrapped child was clicked on, return.
+      if (!el || event.target === el) {
+        return;
+      }
+      el.focus();
+      el.click();
+    }
+
+    _currentLabelTarget() {
+      // Query inside of the current scope. This could either be a shadow root
+      // or the main document.
+      let scope = this.getRootNode();
+      return scope.querySelector(`[aria-labelledby="${this.id}"]`);
+    }
+
+    /**
+     * If there is a for property, return the element with that id.
+     * Else, return the explicitly labeled child.
+     * Else, return the first element child (assume it's implicitly labeled).
+     */
+    _findTarget() {
+      if (this.for) {
+        // external target
+        let scope = this.getRootNode();
+        return scope.getElementById(this.for);
+      }
+
+      let slottedChildren = this._slot.assignedNodes({flatten: true});
+      let el;
+      slottedChildren.forEach(child => {
+        // Ignore text nodes
+        if (child.nodeType === Node.TEXT_NODE) {
+          return;
+        }
+        // Find the first element child. This is the implicit target
+        // of the label.
+        if (!el) {
```

Comment:
Still not very happy about this bit. I see your point but I think looping twice is a small price to pay for the readability here (and the number of children should be <10 on average, no?).

I won’t block on this, it just seems a bit weird (and harder to understand) to give the first iteration a special code branch.

Reply:
Changing to your previously pitched approach to make the code easier to understand.